heart/heartbeat.py

```diff
@@ -124,40 +137,55 @@ def update(self, et, ct):
 		self.push_buf(ct)
 		if self.sockfd:
 			self.reactor.push(ct + 0.1, self.update)
-
+	
 	def get_file_index(self):
+		# for the functional programming noobs:
+		# map takes a function and applies it it every elem in a iterable
 		l = self.server.get_ib_fields()
-		s = "<html>\n<head>\n"
-		s += "<title>Iceball Server List</title>\n"
-		s += "<link href=\"//netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css\" rel=\"stylesheet\">\n"
-		s += "<link rel=\"stylesheet\" type=\"text/css\" href=\"style.css\">\n"
-		s += "</head>\n<body>\n"
-		s += "<div class=\"content\">\n"
-		s += "<h1>Iceball Server List</h1>\n"
-		s += "<table class=\"table table-bordered table-striped table-hover\">\n"
-		s += "<thead>"
-		s += "<th>Address</th>"
-		s += "<th>Port</th>"
-		s += "<th>Name</th>"
-		s += "<th>Version</th>"
-		s += "<th>Players</th>"
-		s += "<th>Mode</th>"
-		s += "<th>Map</th>"
-		s += "</thead>\n"
-		for d in l:
-			s += "<tr>"
-			s += "<td>" + sanestr(d["address"]) + "</td>"
-			s += "<td>" + sanestr(d["port"]) + "</td>"
-			s += "<td>" + "<a href=\"iceball://" + sanestr(d["address"]) + ":" + sanestr(d["port"]) + "\">" + sanestr(d["name"]) + "</a>" + "</td>"
-			s += "<td>" + sanestr(d["version"]) + "</td>"
-			s += "<td>" + sanestr(d["players_current"]) + " / " + sanestr(d["players_max"]) + "</td>"
-			s += "<td>" + sanestr(d["mode"]) + "</td>"
-			s += "<td>" + sanestr(d["map"]) + "</td>"
-			s += "</tr>\n"
-		s += "</table>\n"
-		s += "</div>\n"
-		s += "</body>\n</html>\n"
-		return "text/html", s.replace("\n","\r\n")
+
+		page_template = """\
+<html>
+	<head>
+		<title>Iceball Server List</title>
+		<link href="//netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css"
+			rel="stylesheet">
+		<link rel="stylesheet" type="text/css" href="style.css">
+	</head>
+	<body>
+		<div class="content">
+			<h1>Iceball Server List</h1>
+			<table class="table table-bordered table-striped table-hover">
+				<thead>
+					<th>Address</th>
+					<th>Port</th>
+					<th>Name</th>
+					<th>Version</th>
+					<th>Players</th>
+					<th>Mode</th>
+					<th>Map</th>
+				</thead>
+			{listing}
+			</table>
+		</div>
+	</body>
+</html>
+		"""
+		listing_template = "<tr>"\
+		"<td>{address}</td>"\
+		"<td>{port}</td>"\
+		"<td><a href=\"iceball://{address}:{port}\">{name}</a></td>"\
+		"<td>{version}</td>"\
+		"<td>{players_current}/{players_max}</td>"\
+		"<td>{mode}</td>"\
+		"<td>{map}</td>"\
+		"</tr>\n"
+		
+		listing_list = [listing_template.format(**self.sanetize_dict_values(d)) for d in l]
```

Comment:
No need to make this a list - you can pass the generator straight into `join()`.